pageserver/src/storage_sync.rs

```diff
@@ -1,75 +1,181 @@
+//! there are in-flight operations.
+//! We keep track of the desired remote state in
+//! [`UploadQueueInitialized::latest_files`] and [`UploadQueueInitialized::latest_metadata`].
+//! It is initialized based on the [`IndexPart`] that was passed during init
+//! and updated with every `schedule_*` function call.
+//! All this is necessary necessary to compute the future [`IndexPart`]s
+//! when scheduling an operation while other operations that also affect the
+//! remote [`IndexPart`] are in flight.
+//!
+//! # Retries & Error Handling
+//!
+//! The client retries operations indefinitely, using exponential back-off.
+//! There is no way to force a retry, i.e., interrupt the back-off.
+//! This could be built easily.
 //!
-//! At pageserver startup, it downloads the index files of every locally-present
-//! tenant, to synchronize the local state with remote storage. Other tenants
-//! that might be present in the remote storage are ignored. When a tenant is
-//! attached to the pageserver, the directory structure for the tenant and all
-//! its timelines are created on local disk, and the timeline index files are
-//! downloaded, to create the local metadata files. The layer files containing
-//! the actual data are downloaded on-demand.
+//! # Cancellation
+//!
+//! The operations execute as plain [`task_mgr`] tasks, scoped to
+//! the client's tenant and timeline.
+//! Dropping the client will drop queued operations but not executing operations.
+//! These will complete unless the `task_mgr` tasks are cancelled using `task_mgr`
+//! APIs, e.g., during pagserver shutdown or tenant detach.
+//!
+//! # Completion
+//!
+//! Once an operation has completed, we update
+//! [`UploadQueueInitialized::last_uploaded_consistent_lsn`] which indicates
+//! to safekeepers that they can delete the WAL up to that LSN.
+//!
+//! The [`RemoteTimelineClient::wait_completion`] method can be used to wait
+//! for all pending operations to complete. It does not prevent more
+//! operations from getting scheduled.
+//!
+//! # Crash Consistency
+//!
+//! We do not persist the upload queue state.
+//! If we drop the client, or crash, all unfinished operations are lost.
+//!
+//! To recover , the following steps need to be taken:
+//! - Retrieve the current remote [`IndexPart`]. This gives us a
+//!   consistent remote state, assuming the user scheduled operations in
+//!   the correct order.
+//! - Initiate upload queue with that [`IndexPart`].
+//! - Reschedule all lost operations by comparing the local filesystem state
+//!   and remote state as per [`IndexPart`]. This is done in
+//!   [`Timeline::setup_timeline`] and [`Timeline::reconcile_with_remote`].
+//!
+//! Note that we currently tolerate leaking files on the remote in the case
+//! of interrupted delete operations where we crash inbetween the index
+//! update that removes the reference to the layer file and the actual deletion
+//! of the layer file.
```

Comment:
We can similarly leak a newly created file

```suggestion
//! Note that if we crash during file deletion between the index update
//! that removes the file from the list of files, and deleting the remote file,
//! the file is leaked in the remote storage. Similarly, if a new file is created
//! and uploaded, but the pageserver dies permantently before updating the
//! remote index file, the new file is leaked in remote storage. We accept and
//! tolerate that for now.
```